File: apps/api/internal/tools/cpf.go
package tools

import (
	"fmt"
	"math/rand"
	"regexp"
)

func cpfCalcDigit(digits []int, weights []int) int {
	sum := 0
	for i, d := range digits {
		sum += d * weights[i]
	}
	rem := sum % 11
	if rem < 2 {
		return 0
	}
	return 11 - rem
}

// allSameDigits returns true when every character in s is identical.
// Go's regexp/RE2 does not support backreferences, so we check manually.
func allSameDigits(s string) bool {
	if len(s) == 0 {
		return false
	}
	first := s[0]
	for i := 1; i < len(s); i++ {
		if s[i] != first {
			return false
		}
	}
	return true
}

func ValidarCPF(value string) bool {
	re := regexp.MustCompile(`\D`)
	digits := re.ReplaceAllString(value, "")
	if len(digits) != 11 {
		return false
	}
	if allSameDigits(digits) {
		return false
	}
	nums := make([]int, 11)
	for i, c := range digits {
		nums[i] = int(c - '0')
	}
	d1 := cpfCalcDigit(nums[:9], []int{10, 9, 8, 7, 6, 5, 4, 3, 2})
	d2 := cpfCalcDigit(nums[:10], []int{11, 10, 9, 8, 7, 6, 5, 4, 3, 2})
	return nums[9] == d1 && nums[10] == d2
}

func GerarCPF() string {
	d := make([]int, 9)
	for i := range d {
		d[i] = rand.Intn(10)
	}
	d1 := cpfCalcDigit(d, []int{10, 9, 8, 7, 6, 5, 4, 3, 2})
	d = append(d, d1)
	d2 := cpfCalcDigit(d, []int{11, 10, 9, 8, 7, 6, 5, 4, 3, 2})
	d = append(d, d2)
	return fmt.Sprintf("%d%d%d.%d%d%d.%d%d%d-%d%d",
		d[0], d[1], d[2], d[3], d[4], d[5], d[6], d[7], d[8], d[9], d[10])
}

File: packages/tools/src/cpf.ts
function calcDigit(digits: number[], weights: number[]): number {
  const sum = digits.reduce((acc, d, i) => acc + d * weights[i], 0)
  const rem = sum % 11
  return rem < 2 ? 0 : 11 - rem
}

export function gerarCPF(): string {
  const d = Array.from({ length: 9 }, () => Math.floor(Math.random() * 10))
  d.push(calcDigit(d, [10, 9, 8, 7, 6, 5, 4, 3, 2]))
  d.push(calcDigit(d, [11, 10, 9, 8, 7, 6, 5, 4, 3, 2]))
  return `${d.slice(0, 3).join('')}.${d.slice(3, 6).join('')}.${d.slice(6, 9).join('')}-${d.slice(9).join('')}`
}

export function validarCPF(value: string): boolean {
  const digits = value.replace(/\D/g, '')
  if (digits.length !== 11) return false
  if (/^(\d)\1{10}$/.test(digits)) return false
  const nums = digits.split('').map(Number)
  const d1 = calcDigit(nums.slice(0, 9), [10, 9, 8, 7, 6, 5, 4, 3, 2])
  const d2 = calcDigit(nums.slice(0, 10), [11, 10, 9, 8, 7, 6, 5, 4, 3, 2])
  return nums[9] === d1 && nums[10] === d2
}
